src/main/scala/viper/gobra/translator/encodings/maps/MathematicalMapEncoding.scala

```diff
@@ -0,0 +1,118 @@
+    case _ => Violation.violation("Unexpected case reached")
+  }
+
+  /**
+    * Translates a type into a Viper type.
+    */
+  override def typ(ctx: Context): in.Type ==> vpr.Type = {
+    case ctx.MathematicalMap(k, v) / m =>
+      m match {
+        case Exclusive => vpr.MapType(ctx.typeEncoding.typ(ctx)(k), ctx.typeEncoding.typ(ctx)(v))
+        case Shared    => vpr.Ref
+      }
+  }
+
+  /**
+    * Encodes an assignment.
+    * The first and second argument is the left-hand side and right-hand side, respectively.
+    * [(e: mmap[K]V)[idx] = rhs] -> [ e = e[idx := rhs] ]
+    */
+  override def assignment(ctx: Context): (in.Assignee, in.Expr, in.Node) ==> CodeWriter[vpr.Stmt] = default(super.assignment(ctx)){
+    case (in.Assignee(in.IndexedExp(base :: ctx.MathematicalMap(_, _), idx) :: _ / Exclusive), rhs, src) =>
+      for {
+        isCompKey <- MapEncoding.checkKeyComparability(idx)(ctx)
+        _ <- assert(isCompKey, comparabilityErrorT) // key must be comparable
+        stmt <- ctx.typeEncoding.assignment(ctx)(in.Assignee(base), in.GhostCollectionUpdate(base, idx, rhs)(src.info), src)
+      } yield stmt
+  }
+
+  /**
+    * Encodes expressions as values that do not occupy some identifiable location in memory.
+    *
+    * To avoid conflicts with other encodings, a leaf encoding for type T should be defined at:
+    * (1) exclusive operations on T, which includes literals and default values
+    *
+    * Most cases are a one-to-one mapping to Viper's sequence operations.
+    * R[ dflt(mmap[K]V°) ] -> EmptyMap([K], [V])
+    * R[ mmapLit(E) ] -> ExplicitMap([E])
+    * R[ (e: mmap[K]V)[idx] ] -> MapLookup([e], [idx])
+    * R[ (e: mmap[K]V)[idx = val] ] -> MapUpdate([e], [idx], [val])
+    * R[ len(e: mmap[K]V) ] -> MapCardinality([e])
+    * R[ keySet(e: mmap[K]V) ] -> MapDomain([e])
+    * R[ valueSet(e: mmap[K]V) ] -> MapRange([e])
+    */
+  override def expr(ctx: Context): in.Expr ==> CodeWriter[vpr.Exp] = {
+    def goE(x: in.Expr): CodeWriter[vpr.Exp] = ctx.expr.translate(x)(ctx)
+
+    default(super.expr(ctx)){
+      case (e: in.DfltVal) :: ctx.MathematicalMap(k, v) / Exclusive =>
+        unit(withSrc(vpr.EmptyMap(ctx.typeEncoding.typ(ctx)(k), ctx.typeEncoding.typ(ctx)(v)), e))
+
+      case (lit: in.MathMapLit) :: ctx.MathematicalMap(_, _) / Exclusive =>
+        val (pos, info, errT) = lit.vprMeta
+        for {
+          mapletList <- sequence(lit.entries.toVector.map {
+            case (key, value) => for {
+              k <- goE(key)
+              v <- goE(value)
+            } yield vpr.Maplet(k, v)(pos, info, errT)
+          })
+        } yield vpr.ExplicitMap(mapletList)(pos, info, errT)
+
+      case n@ in.IndexedExp(e :: ctx.MathematicalMap(_, _), idx) =>
+        val (pos, info, errT) = n.vprMeta
+        for {
+          isCompKey <- MapEncoding.checkKeyComparability(idx)(ctx)
+          _ <- assert(isCompKey, isCompKey, comparabilityErrorT)(ctx)
```

Comment:
Check previous comment

Reply:
Why does `isCompKey` occur twice as argument?

Reply:
this is just a hack so that I can fulfill the new signature of `assert`:
```scala
    /* Can be used in expressions. */
    def assert(cond: vpr.Exp, exp: vpr.Exp, reasonT: (Source.Verifier.Info, ErrorReason) => VerificationError)(ctx: Context): Writer[vpr.Exp] = {
      // In the future, this might do something more sophisticated
      val (res, errT) = ctx.condition.assert(cond, exp, reasonT)
      errorT(errT).map(_ => res)
    }
```
that Felix introduced in a previous PR. Introducing a variant of this method that can be used in expressions and that only expects the condition to check and the error transformer introduces ambiguities when calling such method due to ambiguous references to overloaded definition